### PR TITLE
Add SSR helper and server-side variant persistence

### PR DESCRIPTION
**Pull request by jasonscharf**
Adds a helper method for rendering + selecting variants during SSR in Next.js

### CODE REVIEW

**Comment by jasonscharf** [comment]
Current dependencies on/for this PR:
* main
  * **PR #20** <a href="https://app.graphite.dev/github/pr/InstantDomain/instant-bandit/20" target="_blank">Graphite</a> 
    * **PR #21** <a href="https://app.graphite.dev/github/pr/InstantDomain/instant-bandit/21" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/InstantDomain/instant-bandit/21?utm_source=stack-comment).

**Review by hartshorne — APPROVED** [review]

File: lib/server/backends/sessions.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ export function getStubSessionsBackend(): SessionsBackend {
         session = sessions[sid!]
         if (!exists(session)) {
           console.warn(`[IB] Missing or expired session '${sid}'`)
+          session = null
```
Comment (hartshorne) [inline lib/server/backends/sessions.ts:26]: `return null`

File: lib/server/backends/sessions.ts [hunk 2]
```diff
@@ -36,5 +37,22 @@ export function getStubSessionsBackend(): SessionsBackend {
       }
       return session
     },
+
+    async markVariantSeen(session: SessionDescriptor, experimentId: string, variantName: string) {
+      let variants = session.variants[experimentId]
+      if (!exists(variants)) {
+        variants = session.variants[experimentId] = []
+      }
```
Comment (hartshorne) [inline lib/server/backends/sessions.ts:45]: maybe `let variants = session.variants[experimentId] || []` instead?

File: lib/server/server.ts [hunk 3]
```diff
@@ -123,15 +123,15 @@ export async function embedProbabilities(req: ValidatedRequest, origSite: Site,
...
 
-      const bucket = variantMetrics.get(variant);
+      const bucket = variantMetrics.get(variant) as Record<constants.DefaultMetrics, number>
 
       // Show raw metrics in dev mode
       if (env.isDev()) {
         (<VariantMeta>variant).metrics = bucket!
       }
 
       exposures[variant.name] = bucket!.exposures ?? 0
```
Comment (hartshorne) [inline lib/server/server.ts:133]: maybe best not to make so much use of `!.`?

**Review by hartshorne — APPROVED** [review]

File: lib/server/backends/sessions.ts (hunk 1 above)
Comment (jasonscharf) [reply lib/server/backends/sessions.ts]: Can't return here, need to create a new blank session below. Removed the null assignment however.

**Pull request merged**